Add carbon emissions admin configuration options

Renni771 opened the pull request:
Addresses #15046. This feature branch adds admin config parameters and admin documentation for carbon emissions reporting. The following admin flags were added:

- `geographical_server_location_code`  an [ISO 3166](https://en.wikipedia.org/wiki/ISO_3166) code specifying the geographical location of the galaxy instance.
- `power_usage_effectiveness`  the PUE value to use in carbon emissions calculations.
- `carbon_emission_estimates`  a feature toggle flag so reporting can be completely disabled when needed.

The following flags are automatically set based on the value of `geographical_server_location_code`:
- `geographical_server_location_name`  the actual name corresponding to `geographical_server_location_code`.
- `carbon_intensity` the carbon intensity value corresponding to the configured location.

The location is picked from a list of supported locations in [ISO 3166](https://en.wikipedia.org/wiki/ISO_3166) format which is documented [over here](https://github.com/galaxyproject/galaxy/blob/6a15748250b0550e1cd28e2078f51296a3c5d52f/doc/source/admin/carbon_emissions.md). This standard allows users to specify both countries and regions within larger countries. If no location is set or the configured location is invalid or unsupported, the config logic defaults to global values. The UI is dynamic and will display the location name, carbon intensity and PUE values. If global values are used, the UI reflects this too. Here's what the UI looks like:

![Screenshot from 2023-06-24 13-48-32](https://github.com/galaxyproject/galaxy/assets/57318507/fdd56099-fba1-48f2-b16d-a134345639bf)

Here's a list of further additions:
- Documentation for how to configure the carbon emissions feature has been added to `doc/source/admin/carbon_emissions.md`.
- The docs and config files all point to where information on where supported locations can be found. So,  instead of specifying the format, we just refer admins to the list of supported values.
- The docs include a link to resources on how to calculate PUE.
- When `geographical_server_location_code` is invalid or unsupported, the global default values are used and the event is logged.

What else changed:
- The flags in `lib/galaxy/config/__init__.py` were refactored to be sorted alphabetically.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Edit any of the aforementioned flags in the `galaxy.yml.sample` file and assess the behaviour on the client.
  2. Note: When setting a location, consult the [list of supported locations](https://github.com/galaxyproject/galaxy/blob/6a15748250b0550e1cd28e2078f51296a3c5d52f/doc/source/admin/carbon_emissions.md) mentioned in the docs.
## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

Renni771 committed:
Add carbon emissions admin configuration options

- Server location can be set as a valid ISO 3166 code. This is used to
  determine the carbon intensity based off of the galaxy instance server
  location
- A non-default power usage effectiveness value can be set
- Carbon emissions reporting can be toggled

Renni771 committed:
Move carbon intensity reference data to backend and add admin docs
- Add `geographical_server_location_code`, `geographical_server_location_name` and `carbon_intensity` flags.
- Alphabetically sort the list of config flags in galaxy init file.
- Add logic to set `geographical_server_location_name` and `carbon_intensity` based off of `geographical_server_location_code`.

Renni771 committed:
Format code

Renni771 committed:
Merge dev

Renni771 committed:
Rebuild config to set `geographical_server_location_code` to its default
value

Renni771 committed:
Fix carbon_intensity.csv formatting

Renni771 committed:
Correctly type power_usage_effectiveness flag in config

Renni771 committed:
Add defaults to carbon intensity and PUE props

Renni771 committed:
Merge branch 'dev' into add-carbon-emissions-admin-configuration-options

Renni771 committed:
Use f-string format for region name string

Renni771 committed:
Fix job metrics heading hierarchy

Renni771 committed:
Move logic to get carbon intensity entry for location into its own module

Renni771 committed:
Replace `yield` with `yield from` for carbon intensity csv file reader

Renni771 committed:
Correctly make carbon_emissions its own module by renaming the init file

Renni771 committed:
Fix typo and use `log.warning` instead of `log.warn`

Renni771 committed:
Add `lib/galaxy/carbon_emissions` module symlink to `packages/app/galaxy/` directory

Renni771 committed:
Add unit tests for `get_carbon_intensity_entry`

Renni771 committed:
Add client side tests for dynamic PUE, carbon intensity and location values

- Simplify the component API by making its consumers pass all necessary props

Renni771 committed:
Run formatter on code

Renni771 committed:
Fix types in config_schema

Renni771 committed:
Merge dev and resolve conflicts

bgruening reviewed (commented):
Really nice @Renni771 thanks a lot!

Anyone from the backend team that wants to have a look?

Renni771 committed:
Fix copy paste error

davelopez commented on lib/galaxy/managers/configuration.py:
```diff
@@ -175,6 +175,10 @@ def _config_is_truthy(item, key, **context):
+            "carbon_emission_estimates": _defaults_to(True),
+            "carbon_intensity": _use_config,
+            "geographical_server_location_name": _use_config,
+            "power_usage_effectiveness": _use_config,
```
Looks like `geographical_server_location_code` doesn't get serialized here.

davelopez commented on lib/galaxy/config/__init__.py:
```diff
@@ -836,6 +840,11 @@ def _process_config(self, kwargs: Dict[str, Any]) -> None:
         else:
             self.version_extra = extra_info
 
+        # Carbon emissions configuration
+        carbon_intensity_entry = get_carbon_intensity_entry(kwargs.get("geographical_server_location_code", ""))
+        self.carbon_intensity = carbon_intensity_entry["carbon_intensity"]
+        self.geographical_server_location_name = carbon_intensity_entry["location_name"]
```
It seems that `carbon_intensity` and `geographical_server_location_name` are always computed here. Therefore I think they should not be part of the `config_schema.yml` since whatever value you set in the `galaxy.yml` config file will be ignored by this. Also, the fewer config options the better :)

davelopez reviewed (commented):
Really cool indeed!

Just a couple of comments below :)

Renni771 committed:
Serialize geographical location code and remove unneeded carbon emissions flags

- `carbon_intensity` and `geographical_server_location_name` are
  automatically computed on startup and are never set by an admin. So
  they don't need to be in the config at all.

davelopez reviewed (approved):
Awesome! Thank you @Renni771!

mvdbeek commented on lib/galaxy/config/__init__.py:
```diff
@@ -34,6 +34,7 @@
 
 import yaml
 
+from galaxy.carbon_emissions import get_carbon_intensity_entry
```
This is breaking the config module

Renni771 replied:
Is this still causing an issue and, if so, how exactly is it breaking the config module?

mvdbeek replied:
it is, the config package can't import from carbon_emissions.

jdavcs replied:
Work on a fix is underway in #17159

Renni771 replied:
Thanks for the update!
